README.md
# oid-node
[![NPM Version](https://img.shields.io/npm/v/@y0x54a/oid-node)](https://www.npmjs.com/package/@y0x54a/oid-node)
[![Build Status](https://github.com/y0x54a/oid-node/workflows/ci/badge.svg?branch=main)](https://github.com/y0x54a/oid-node/actions)
[![codecov](https://codecov.io/gh/y0x54a/oid-node/branch/main/graph/badge.svg?token=58AFJJERGB)](https://codecov.io/gh/y0x54a/oid-node)

## Installing
```sh
npm i @y0x54a/oid-node
```

## Example
```js
const {Oid453, Oid563} = require('@y0x54a/oid-node');
```

```js
Oid453.create();
// 12-byte

Oid563.create();
// 14-byte
```

```js
const oid = new Oid453('00112233445566778899aabb');

oid.getId();
// 00112233445566778899aabb

oid.getTimestamp();
// 1122867

oid.getRandom();
// 293490554760

oid.getIndex();
// 10070715

oid.toString();
// 00112233445566778899aabb

Oid453.validate(oid);
// true

Oid453.validate(oid.getId());
// true

Oid453.validate('00112233445566778899AABB');
// false

Oid563.validate(oid);
// false
```

```js
const oid1 = Oid453.create(1122867);
const oid2 = Oid453.create(oid1);

oid1.getId() === oid2.getId();
// true

oid1.getTimestamp();
// 1122867

oid2.getTimestamp();
// 1122867

Oid453.generate(1122867);
// 00112233...
```

```js
const oid = new Oid563('00112233445566778899aabbccdd');

oid.getId();
// 00112233445566778899aabbccdd

oid.getTimestamp();
// 287454020

oid.getRandom();
// 93898580466090

oid.getIndex();
// 12307677

oid.toString();
// 00112233445566778899aabbccdd

Oid563.validate(oid);
// true

Oid563.validate(oid.getId());
// true

Oid563.validate('00112233445566778899AABBCCDD');
// false

Oid453.validate(oid);
// false
```

```js
const oid1 = Oid563.create(287454020);
const oid2 = Oid563.create(oid1);

oid1.getId() === oid2.getId();
// true

oid1.getTimestamp();
// 287454020

oid2.getTimestamp();
// 287454020

Oid563.generate(287454020);
// 0011223344...
```

## API

- ### Oid

- ### Oid453

  - **Methods**

  - `constructor(id: string | Oid)`

  - `getId(): string`

  - `getTimestamp(): number`

  - `getRandom(): number`

  - `getIndex(): number`

  - `toString(): string`

  - **Static Methods**

  - `create(id?: string | Oid | number): Oid453`

  - `validate(id: string | Oid): boolean`

  - `generate(timestamp: number): string`

- ### Oid563

  - **Methods**

  - `constructor(id: string | Oid)`

  - `getId(): string`

  - `getTimestamp(): number`

  - `getRandom(): number`

  - `getIndex(): number`

  - `toString(): string`

  - **Static Methods**

  - `create(id?: string | Oid | number): Oid563`

  - `validate(id: string | Oid): boolean`

  - `generate(timestamp: number): string`

CHANGELOG.md
# 0.1.0 - 2022-07-14

- Release

test/jest.config.js
module.exports = {
  rootDir: '../',
  testEnvironment: 'node',
  coverageProvider: 'v8',
  coverageReporters: [['json', {file: 'coverage-json.json'}], 'text'],
  coverageDirectory: 'build/coverage/',
  collectCoverageFrom: ['src/**/*.js'],
  verbose: true
};

test/src/oid.test.js
'use strict';

const {Oid} = require('../../src');
const {validOid453Inputs} = require('./resources');

describe('Oid', () => {
  test('constructor', () => {
    const input = validOid453Inputs[0];
    const oid1 = new Oid(input[0]);
    const oid2 = new Oid(oid1);
    expect(oid2._id).toBe(oid1._id);
  });

  test('getId', () => {
    const input = validOid453Inputs[0];
    const oid = new Oid(input[0]);
    expect(oid._id).toBe(input[0]);
    expect(oid.getId()).toBe(input[0]);
  });

  test('getTimestamp', () => {
    const input = validOid453Inputs[0];
    const oid = new Oid(input[0]);
    expect(oid._timestamp).toBeNull();
    expect(oid.getTimestamp()).toBe(input[1]);
    expect(oid._timestamp).not.toBeNull();
  });

  test('getRandom', () => {
    const input = validOid453Inputs[0];
    const oid = new Oid(input[0]);
    expect(oid._random).toBeNull();
    expect(oid.getRandom()).toBe(input[1]);
    expect(oid._random).not.toBeNull();
  });

  test('getIndex', () => {
    const input = validOid453Inputs[0];
    const oid = new Oid(input[0]);
    expect(oid._index).toBeNull();
    expect(oid.getIndex()).toBe(input[1]);
    expect(oid._index).not.toBeNull();
  });

  test('toString', () => {
    const input = validOid453Inputs[0];
    const oid = new Oid(input[0]);
    expect(oid.toString()).toBe(input[0]);
  });
});

test/src/resources.js
'use strict';

const validOid453Inputs = [
  ['000000000000000000000000', 0x00000000, 0x0000000000, 0x000000],
  ['00000000ffffffffffffffff', 0x00000000, 0xffffffffff, 0xffffff],
  ['ffffffffffffffffffffffff', 0xffffffff, 0xffffffffff, 0xffffff],
  ['00112233445566778899aabb', 0x00112233, 0x4455667788, 0x99aabb]
];

const invalidOid453Inputs = [
  '0000000000000000000000',
  'ffffffffffffffffffffffffff',
  'FF0000000000000000000000',
  'ffffffffffffffffffffffFF',
  'aabbccddeeffgghhiijjkkll',
  '00112233445566778899aabbccdd'
];

const validOid563Inputs = [
  ['0000000000000000000000000000', 0x0000000000, 0x000000000000, 0x000000],
  ['0000000000ffffffffffffffffff', 0x0000000000, 0xffffffffffff, 0xffffff],
  ['ffffffffffffffffffffffffffff', 0xffffffffff, 0xffffffffffff, 0xffffff],
  ['00112233445566778899aabbccdd', 0x0011223344, 0x5566778899aa, 0xbbccdd]
];

const invalidOid563Inputs = [
  '00000000000000000000000000',
  'ffffffffffffffffffffffffffffff',
  'FF00000000000000000000000000',
  'ffffffffffffffffffffffffffFF',
  'aabbccddeeffgghhiijjkkllmmnn',
  '00112233445566778899aabb'
];

module.exports = {
  validOid453Inputs,
  invalidOid453Inputs,
  validOid563Inputs,
  invalidOid563Inputs
};

src/oid.js
'use strict';

class Oid
{
  /**
   * @param {(string|Oid)} id
   */
  constructor(id){
    if (id instanceof Oid) {
      id = id.getId();
    }
    this._id = id;
    this._timestamp = null;
    this._random = null;
    this._index = null;
  }

  /**
   * @returns {string}
   */
  getId(){
    return this._id;
  }

  /**
   * @returns {number}
   */
  getTimestamp(){
    if (this._timestamp === null) {
      const length = this._id.length;
      this._timestamp = parseInt(this._id.substring(0, length / 2 - 4), 16);
    }
    return this._timestamp;
  }

  /**
   * @returns {number}
   */
  getRandom(){
    if (this._random === null) {
      const length = this._id.length;
      this._random = parseInt(this._id.substring(length / 2 - 4, length - 6), 16);
    }
    return this._random;
  }

  /**
   * @returns {number}
   */
  getIndex(){
    if (this._index === null) {
      const length = this._id.length;
      this._index = parseInt(this._id.substring(length - 6, length), 16);
    }
    return this._index;
  }

  /**
   * @returns {string}
   */
  toString(){
    return this._id;
  }
}

module.exports = Oid;

src/utils.js
'use strict';

const crypto = require('crypto');
const hexmap = [
  '00','01','02','03','04','05','06','07','08','09','0a','0b','0c','0d','0e','0f',
  '10','11','12','13','14','15','16','17','18','19','1a','1b','1c','1d','1e','1f',
  '20','21','22','23','24','25','26','27','28','29','2a','2b','2c','2d','2e','2f',
  '30','31','32','33','34','35','36','37','38','39','3a','3b','3c','3d','3e','3f',
  '40','41','42','43','44','45','46','47','48','49','4a','4b','4c','4d','4e','4f',
  '50','51','52','53','54','55','56','57','58','59','5a','5b','5c','5d','5e','5f',
  '60','61','62','63','64','65','66','67','68','69','6a','6b','6c','6d','6e','6f',
  '70','71','72','73','74','75','76','77','78','79','7a','7b','7c','7d','7e','7f',
  '80','81','82','83','84','85','86','87','88','89','8a','8b','8c','8d','8e','8f',
  '90','91','92','93','94','95','96','97','98','99','9a','9b','9c','9d','9e','9f',
  'a0','a1','a2','a3','a4','a5','a6','a7','a8','a9','aa','ab','ac','ad','ae','af',
  'b0','b1','b2','b3','b4','b5','b6','b7','b8','b9','ba','bb','bc','bd','be','bf',
  'c0','c1','c2','c3','c4','c5','c6','c7','c8','c9','ca','cb','cc','cd','ce','cf',
  'd0','d1','d2','d3','d4','d5','d6','d7','d8','d9','da','db','dc','dd','de','df',
  'e0','e1','e2','e3','e4','e5','e6','e7','e8','e9','ea','eb','ec','ed','ee','ef',
  'f0','f1','f2','f3','f4','f5','f6','f7','f8','f9','fa','fb','fc','fd','fe','ff'
];

/**
 * @param {number} size
 * @returns {Buffer}
 */
function createRandom(size){
  return crypto.randomBytes(size);
}

/**
 * @param {number} size
 * @returns {function}
 */
function createSequence(size){
  let sequence = ~~(Math.random() * size);
  return () => (sequence = (sequence + 1) % size);
}

/**
 * @returns {number}
 */
function createTimestamp(){
  return ~~(Date.now() / 1e3);
}

module.exports = {
  hexmap,
  createRandom,
  createSequence,
  createTimestamp
};

src/oid453.js
'use strict';

const Oid = require('./oid');
const {hexmap, createRandom, createSequence, createTimestamp} = require('./utils');
const sequence = createSequence(0xffffff);
const random = createRandom(5);

class Oid453 extends Oid
{
  /**
   * @param {(string|Oid)} id
   * @throws {TypeError}
   */
  constructor(id){
    if (!Oid453.validate(id)) {
      throw new TypeError('Invalid ID');
    }
    super(id);
  }

  /**
   * @param {(string|Oid|number)} [id]
   * @returns {Oid453}
   */
  static create(id){
    if (id === void 0) {
      id = Oid453.generate(createTimestamp());
    } else if (typeof id === 'number') {
      id = Oid453.generate(id);
    }
    return new Oid453(id);
  }

  /**
   * @param {(string|Oid)} id
   * @returns {boolean}
   */
  static validate(id){
    if (id instanceof Oid) {
      id = id.getId();
    }
    return /^[0-9a-f]{24}$/.test(id);
  }

  /**
   * @param {number} timestamp
   * @returns {string}
   */
  static generate(timestamp){
    let id = '';
    const index = sequence();
    /* 4-byte timestamp */
    id += hexmap[(timestamp >> 24) & 0xff];
    id += hexmap[(timestamp >> 16) & 0xff];
    id += hexmap[(timestamp >> 8) & 0xff];
    id += hexmap[timestamp & 0xff];
    /* 5-byte random */
    id += hexmap[random[0]];
    id += hexmap[random[1]];
    id += hexmap[random[2]];
    id += hexmap[random[3]];
    id += hexmap[random[4]];
    /* 3-byte index */
    id += hexmap[(index >> 16) & 0xff];
    id += hexmap[(index >> 8) & 0xff];
    id += hexmap[index & 0xff];
    return id;
  }
}

module.exports = Oid453;

src/oid563.js
'use strict';

const Oid = require('./oid');
const {hexmap, createRandom, createSequence, createTimestamp} = require('./utils');
const sequence = createSequence(0xffffff);
const random = createRandom(6);

class Oid563 extends Oid
{
  /**
   * @param {(string|Oid)} id
   * @throws {TypeError}
   */
  constructor(id){
    if (!Oid563.validate(id)) {
      throw new TypeError('Invalid ID');
    }
    super(id);
  }

  /**
   * @param {(string|Oid|number)} [id]
   * @returns {Oid563}
   */
  static create(id){
    if (id === void 0) {
      id = Oid563.generate(createTimestamp());
    } else if (typeof id === 'number') {
      id = Oid563.generate(id);
    }
    return new Oid563(id);
  }

  /**
   * @param {(string|Oid)} id
   * @returns {boolean}
   */
  static validate(id){
    if (id instanceof Oid) {
      id = id.getId();
    }
    return /^[0-9a-f]{28}$/.test(id);
  }

  /**
   * @param {number} timestamp
   * @returns {string}
   */
  static generate(timestamp){
    let id = '';
    const index = sequence();
    /* 5-byte timestamp */
    id += hexmap[~~(timestamp / (0xffffffff + 1)) & 0xff];
    id += hexmap[(timestamp >> 24) & 0xff];
    id += hexmap[(timestamp >> 16) & 0xff];
    id += hexmap[(timestamp >> 8) & 0xff];
    id += hexmap[timestamp & 0xff];
    /* 6-byte random */
    id += hexmap[random[0]];
    id += hexmap[random[1]];
    id += hexmap[random[2]];
    id += hexmap[random[3]];
    id += hexmap[random[4]];
    id += hexmap[random[5]];
    /* 3-byte index */
    id += hexmap[(index >> 16) & 0xff];
    id += hexmap[(index >> 8) & 0xff];
    id += hexmap[index & 0xff];
    return id;
  }
}

module.exports = Oid563;